### PR TITLE
add go1.11 module support and update lsync version

### PR DESCRIPTION
**Pull request by wlan0**
Found a bug in https://github.com/minio/minio/commit/ce870466ff5db524e5f6c54b97eb0b26d9fcdf46 where the vendor directory was updated, but not the upstream project `github.com/minio/lsync`

On further investigation found that the api `GetLock(id, source string, timeout time.Time) bool` only applies for `github.com/minio/dsync` project, and not `github.com/minio/lsync` project. So, i've updated the implementation to use the upstream project appropriately. 

This bug only showed up because Im using go-1.11.4. For other using go-1.11.x or greater this issue would show up. In order to prevent this, I added go.mod file to ensure that go-1.11.4 users get a consistent version of the dependenices as a part of the same PR.

### CODE REVIEW

**Review by harshavardhana — CHANGES_REQUESTED** [review]
This may not be a correct change, we should have an RWLockerSync interface and make lsync.LRWMutex work the same way. 

Because we would need to List locks in the same manner for Lsync style locks as well.

**Comment by wlan0** [comment]
You can list locks on sync.Locker interface. 

lsync.LRWMutex does not use `id` or `source`. Why should we change a project to satisfy an interface in a depending project?

**Comment by harshavardhana** [comment]
> lsync.LRWMutex does not use `id` or `source`. Why should we change a project to satisfy an interface in a depending project?

id, source are required even in case of lsync just like dsync. The implementation here to use multiple styles is wrong, lsync and dsync are meant to comply with a common interface - so any feature added to dsync such as source should be available in lsync as well.

**Comment by wlan0** [comment]
> id, source are required even in case of lsync just like dsync. The implementation here to use multiple > styles is wrong, lsync and dsync are meant to comply with a common interface - so any feature 
> added to dsync such as source should be available in lsync as well.

since I can't tell how lsync uses `id` and `source`, and right now this build is failing, I'm just going to make this change locally, and let `lsync` be fixed by people who know it better. Closing this PR

**Comment by harshavardhana** [comment]
> since I can't tell how lsync uses `id` and `source`, and right now this build is failing, I'm just going to make this change locally, and let `lsync` be fixed by people who know it better. Closing this PR

Let's keep the PR open, we need to add `go modules` anyways. Let's fix the lsync and then rebase it here.

**Comment by wlan0** [comment]
ok, lets chat about lsync tomorrow in office and go from there

File: cmd/namespace-lock.go [hunk 1]
```diff
@@ -120,8 +112,28 @@ type nsParam struct {
...
+	LRWMutexLock *lsync.LRWMutex
+	ref          uint
+}
+
+func (n *nsLock) Unlock() {
+	if n.DRWMutexLock != nil {
+		n.DRWMutexLock.Unlock()
+	}
+	if n.LRWMutexLock != nil {
+		n.LRWMutexLock.Unlock()
```
Comment (harshavardhana) [inline cmd/namespace-lock.go]: Can you update this PR to using latest lsync? with `go.mod` ?

Comment (wlan0) [reply cmd/namespace-lock.go]: done

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |

File: Makefile [hunk 2]
```diff
@@ -19,6 +19,22 @@ getdeps:
 	@echo "Installing misspell" && go get -u github.com/client9/misspell/cmd/misspell
 	@echo "Installing ineffassign" && go get -u github.com/gordonklaus/ineffassign
 
+prunevendor:
+	@find vendor/ -name '.editorconfig' | xargs -I{} rm -v {}
```
Comment (harshavardhana) [inline Makefile:23]: Add this to CONTRIBUTING.md as well to deprecate `pkg-add`, `pkg-remove` etc. remove them from Makefile. We need to update the CONTRIBUTING.md respectively.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: Makefile (hunk 2 above)
Comment (wlan0) [reply Makefile:23]: Will do.

**Pull request closed without merging**